velox/functions/sparksql/AtLeastNNonNulls.h

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/functions/Macros.h"
+
+namespace facebook::velox::functions::sparksql {
+
+template <typename T>
+struct AtLeastNNonNullsFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void callNullable(
+      out_type<bool>& out,
+      const arg_type<int32_t>* n,
+      const arg_type<Variadic<Any>>* inputs) {
+    out = false;
+    int32_t result = 0;
+    int32_t expectedNum = *n;
```

Comment:
what if n is null? and add the test.

Reply:
n here should not be null, In Spark, it's not a expression, just a literal. 
https://github.com/apache/spark/blob/110b558570176d5d5ee5ba85bb071fd66a94a7a0/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala#L428C28-L428C35

The thing here is the `intialize` method can't be added with `Variadic` type arguments, if we add some check to make sure it's not null, the check would happens for every row.
https://github.com/facebookincubator/velox/blob/8e0181ef723854b438b8e56d72bf87cb0cb51333/velox/core/SimpleFunctionMetadata.h#L836

Reply:
OK, let add the constraints in document, and VELOX_DCHEK to check the `n`, it will only make effect in debug version.

Reply:
What do you think? @mbasmanova Should we add this function to fuzzer test black list or ignore the performance issue to use `VELOX_USER_CHECK_NOT_NULL`?

Reply:
Some background, this is a Spark internal function whose argument `n` is Int, so it won't be null, but we cannot check in initialize because variadicargs

Reply:
@jinchengchenghh This function seems rather expensive, hence, I wouldn't expect much perf hit from using a null check. Are you seeing significant performance difference with and without the check?

That said, I wonder if this function needs to be a special form (similar to coalesce). I would imagine this function stops checking the arguments once if found 'n' non-NULL / non-NaN values. Hence, I'd expect it may not evaluate the remaining arguments. Is this how it works in Spark? If so, it would need to be defined as a special form so it can control whether all arguments are evaluated or only some.

Reply:
Right. Looks like Spark stops evaluating arguments if reaching the expected value.

https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala#L438

Reply:
Yes, I agree, as this discussion https://github.com/facebookincubator/velox/pull/10508/files#r1689090578 , we have ended up the loop but all the argument has evaluated.
I read coalesce logic,
```
for (int i = 0; i < inputs_.size(); i++) {
    inputs_[i]->eval(*activeRows, context, result);
```
This can help to not evaluate remaining argument.
Thanks for your suggestion.
Can you help change it?
@zhli1142015

Reply:
I see, let me check.

Reply:
I changed to implement this with special form, please help review.
Thanks.